File: src/part4/ticket_service.ts
import * as restate from "@restatedev/restate-sdk";
import {
  Ticket,
  TicketService as ITicketService,
} from "../generated/proto/example";
import { BoolValue } from "../generated/proto/google/protobuf/wrappers";
import { Empty } from "../generated/proto/google/protobuf/empty";

export class TicketService implements ITicketService {
  async reserve(request: Ticket): Promise<BoolValue> {
    const ctx = restate.useContext(this);

    const ticketStatus = await ctx.get<string>("ticket_status");

    if (!ticketStatus || ticketStatus === "AVAILABLE") {
      ctx.set("ticket_status", "RESERVED");
      return BoolValue.create({ value: true });
    } else {
      return BoolValue.create({ value: false });
    }
  }

  async unreserve(request: Ticket): Promise<Empty> {
    const ctx = restate.useContext(this);

    ctx.set("ticket_status", "AVAILABLE");

    return {};
  }

  async markAsSold(request: Ticket): Promise<Empty> {
    const ctx = restate.useContext(this);

    ctx.set("ticket_status", "SOLD");

    return {};
  }
}

File: java/src/main/java/dev/restate/tour/part4/TicketService.java
package dev.restate.tour.part4;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.protobuf.BoolValue;
import com.google.protobuf.Empty;
import dev.restate.sdk.blocking.RestateBlockingService;
import dev.restate.sdk.blocking.RestateContext;
import dev.restate.sdk.core.StateKey;
import dev.restate.sdk.core.serde.jackson.JacksonSerde;
import dev.restate.tour.auxiliary.TicketStatus;
import dev.restate.tour.generated.TicketServiceGrpc;
import dev.restate.tour.generated.Tour.*;
import io.grpc.stub.StreamObserver;

public class TicketService extends TicketServiceGrpc.TicketServiceImplBase implements RestateBlockingService {

    public static final StateKey<TicketStatus> STATE_KEY = StateKey.of("status", JacksonSerde.typeRef(new TypeReference<>() {}));

    @Override
    public void reserve(Ticket request, StreamObserver<BoolValue> responseObserver) {
        RestateContext ctx = restateContext();

        var status = ctx.get(STATE_KEY).orElse(TicketStatus.Available);

        if(status.equals(TicketStatus.Available)){
            ctx.set(STATE_KEY, TicketStatus.Reserved);
            responseObserver.onNext(BoolValue.of(true));
        } else {
            responseObserver.onNext(BoolValue.of(false));
        }
        responseObserver.onCompleted();
    }

    @Override
    public void unreserve(Ticket request, StreamObserver<Empty> responseObserver) {
        RestateContext ctx = restateContext();

        var status = ctx.get(STATE_KEY).orElse(TicketStatus.Available);

        if(!status.equals(TicketStatus.Sold)){
            ctx.clear(STATE_KEY);
        }
        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }

    @Override
    public void markAsSold(Ticket request, StreamObserver<Empty> responseObserver) {
        RestateContext ctx = restateContext();

        var status = ctx.get(STATE_KEY).orElse(TicketStatus.Available);

        if(status.equals(TicketStatus.Reserved)){
            ctx.set(STATE_KEY, TicketStatus.Sold);
        }
        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }
}
